### PR TITLE
Implement step based checkpointing

### PR DESCRIPTION
**Pull request by joecummings**
- **Add helper functions for pruning old checkpoints**
- **Step based checkpointing and tests**

#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Closes #2105. This is a widely requested feature that allows users to have greater control over checkpointing frequency in torchtune. 

TODO: Add commentary on design decisions. Acknowledge spaghetti code. Beg forgiveness.

#### Changelog
* Update ``FullModelHFCheckpointer`` to accept a step parameter when saving a checkpoint. Use that step to designate the checkpoint folder name. Keep ``epoch_{}`` as a fall-back for BC.
* Add helper function ``prune_surplus_checkpoints`` that removes extra checkpoints in order to free up space for the user. This is controlled by the ``keep_n_last_checkpoints``.
* Add helper function ``get_all_checkpoints_in_dir`` that scans a directory (meant for the output_dir) and returns all checkpoint directories.
* Add tests for corresponding helper functions
* Modify the ``full_finetune_single_device.py`` recipe to utilize step-based checkpointing.

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [x] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

**Ran single device recipe w/ step-based checkpointing**
*Evidence of succesful run*

*Evidence of correct number of checkpoints being saved*

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [x] I have added an example to docs or docstrings

### CODE REVIEW

**Comment by bogdansalyp** [comment]
- `recipe_state` is still saved to `${output_dir}`, not `${output_dir}/step_XXX`
- `resume_from_checkpoint` logic should be updated
  - RN it looks for `${output_dir}` root, not `step_XXX`
  - maybe replace top level `cfg.resume_from_checkpoint` to have `cfg.checkpointer.resume_from` which is either "latest" (default) or the path to the checkpoint to resume from. Or separate mutually exclusive `resume_from: /path/` and `resume_from_latest: True`
  - offtopic but `cfg.resume_from_checkpoint` is mentioned in code as deprecated and replaced by `should_load_recipe_state` but de facto `resume_from_checkpoint` is mandatory and `should_load_recipe_state` doesn't work
- lr schedulers don't work when resuming from `recipe_state`
- step and epoch are recovered properly from `recipe_state` but not used to update `current_step` - as a result logging breaks and

File: recipes/full_finetune_single_device.py [hunk 1]
```diff
@@ -627,14 +663,20 @@ def train(self) -> None:
                             log_dict["grad_norm"] = grad_norm
                         self._metric_logger.log_dict(log_dict, step=self.global_step)
 
+                    # Save checkpoint if specified by user
+                    if self.global_step % self.save_every_n_steps == 0:
+                        self.save_checkpoint(
+                            epoch=curr_epoch, step=self.global_step, full_tensors=False
```
Comment (felipemello1) [inline recipes/full_finetune_single_device.py:669]: var "step" is missing in distributed recipe

**Comment by bogdansalyp** [comment]
rebased to #2869
